neat_ml/opencv/detection.py

```diff
@@ -0,0 +1,210 @@
+from pathlib import Path
+from typing import Sequence
+
+import cv2
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import skimage.color
+from joblib import Memory
+from tqdm.auto import tqdm
+
+
+__all__: Sequence[str] = [
+    "collect_tiff_paths", 
+    "run_opencv"
+]
+
+def collect_tiff_paths(
+    img_dir: str | Path
+) -> list[str]:
+    """
+    Recursively locate every TIFF image beneath a given directory.
+
+    Parameters
+    ----------
+    img_dir : str or Path
+        Root directory to search for files.
+
+    Returns
+    -------
+    list[str]
+        List of absolute file paths for each .tiff file found.
+    """
+    img_path = Path(img_dir).expanduser().resolve()
+    return [str(Path(p).resolve()) for p in img_path.glob("**/*.tiff")]
+
+
+def _detect_single_image(
+    img_path: str
+) -> tuple[int, float, pd.DataFrame]:
+    """
+    Detect bubbles in a single image using OpenCV's SimpleBlobDetector.
+
+    Parameters
+    ----------
+    img_path : str
+        Absolute file path to the image to process.
+
+    Returns
+    -------
+    tuple[int, float, pd.DataFrame]
+        num_blobs : int
+            Number of blobs detected in the image.
+        median_radius : float
+            Median radius of detected blobs (NaN if none).
+        bubble_data : pd.DataFrame
+            DataFrame with one row per detected blob, each containing:
+            - 'bubble_number' (int)
+            - 'center' (tuple[float, float])
+            - 'radius' (float)
+            - 'area' (float)
+            - 'bbox' (tuple[int, int, int, int])
+    """
+    image = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
+    if image is None:
+        raise FileNotFoundError(f"Unable to read image file: {img_path}")
+
+    # the parameters for ``SimpleBlobDetector`` were determined manually
+    # by hand-tuning via visual inspection NOT by using a standardized
+    # hyperparameter optimization method (see: issue #13)
```

Comment:
have you updated the manuscript with a similar admonition if appropriate? If there's no longer any mention of OpenCV in the main paper, this may belong in the SI.